### DISCUSSION TITLE
Managed representation of actors and components

### DISCUSSION THREAD

**Discussion by nowakpe** [nxrighthere/UnrealCLR#40] [Ideas]
Hi, I am trying to use this plugin together with Svelto ECS framework.
I think for these kinds of usages would be nice to have exposed pointers and static methods.
I mean for example Pointer property and static methods from Actor class. 
So it is not necessary to implement some id mappings, as I want to use only structures in ECS.
I have already modified the Actor class for test purposes but would be nice to have it in general.
It is also possible to create struct types like struct ActorStruct or something, but it would be probably a lot of work.

What do you think about that? :-)

**Comment by nxrighthere** [comment]
It doesn't carry any benefits since native objects are not contiguous in memory and scattered all over the place. Exposing pointers directly without encapsulating them into intermediate abstractions for type safety is the way to maintenance hell.

**Comment by nxrighthere** [comment]
Some resources about ECS in Unreal Engine:
 1. https://www.youtube.com/watch?v=QlKXPBFh5BM
 2. https://forums.unrealengine.com/development-discussion/c-gameplay-programming/1449913-implementing-ecs-architecture-in-ue4-giant-space-battle
 3. https://github.com/skypjack/entt/issues/61

**Reply by nowakpe** [reply]
Yep I know all of that, but I can have a simple struct component containing the actual pointer of Actor. And than simply call static method with this pointer, like you are doing that internally. So I don't have to jump even more using some Dictionary<int id,Actor>.
It is true that you are losing safety in the framework. Maybe those static methods can be exposed through ActorInterop class or something, so it is not exposed directly through Actor :-D And then it will be the responsibility of the user if he will use it or not...

**Reply by nxrighthere** [reply]
It's not only about losing safety, but also about losing functionality from inheritance hierarchy unless you start to implement interfaces using generic objects.

If you want to do that, the plugin allows creating an entirely custom framework by only copying `Codegen.cs` to your custom `UnrealEngine.Framework` project implementation and using all internal stuff as you want.
